### PR TITLE
Issue/#567 new map pools

### LINKED ISSUE #567: Add support for new map pool tables
See https://github.com/FAForever/db/pull/214.

Each matchmaker queue can have any number of map pool associated with it. These map pools will be drawn from depending on player rating.

### Added tables
- `map_pool`
- `map_pool_map_version`
- `matchmaker_queue_map_pool`

### PR DESCRIPTION
**Pull request by Askaholic**
Needs database version 82!

Add support for new `map_pool` and `matchmaker_queue` tables. If no entry for `ladder1v1` exists, then fall back to using a hardcoded `ladder1v1` configuration and pull maps from the old `ladder_map` table.

Notes:
- If a queue has no map pools then all games will fail to start
- Map pool is determined by lowest player rating in the game

Also includes some test cleanup:
- Merge `test-data.sql` file
- Make `player_factory` fixture "session" scoped
- Rename `p` to override `player_factory` in test_matchmaker_queue_algorithm.py

Closes #567

### CODE REVIEW

File: server/ladder_service.py [hunk 1]
```diff
@@ -37,19 +41,103 @@ def __init__(
         self._informed_players: Set[Player] = set()
         self.game_service = game_service
 
+        # Fallback legacy map pool
+        self.ladder_1v1_map_pool = MapPool(0, "ladder1v1")
         # Hardcoded here until it needs to be dynamic
```
Comment (Askaholic) [inline server/ladder_service.py:46]: Delete

File: server/ladder_service.py [hunk 2]
```diff
@@ -168,7 +256,22 @@ async def start_game(self, host: Player, guest: Player):
...
             guest.state = PlayerState.JOINING
 
-            (map_id, map_name, map_path) = await self.choose_map([host, guest])
+            played_map_ids = list(flatten(
+                await asyncio.gather(*[
+                    self.get_game_history(player, "ladder1v1")
+                    for player in [host, guest]
+                ])
+            ))
+            rating = max(
```
Comment (Askaholic) [inline server/ladder_service.py:265]:
```suggestion
            rating = min(
```

File: server/ladder_service.py [hunk 3]
```diff
@@ -168,7 +256,22 @@ async def start_game(self, host: Player, guest: Player):
             host.state = PlayerState.HOSTING
             guest.state = PlayerState.JOINING
 
-            (map_id, map_name, map_path) = await self.choose_map([host, guest])
+            played_map_ids = list(flatten(
+                await asyncio.gather(*[
+                    self.get_game_history(player, "ladder1v1")
```
Comment (Askaholic) [inline server/ladder_service.py:261]: Reuse a single db connection .Maybe executemany?

File: tests/unit_tests/test_ladder_service.py [hunk 4]
```diff
@@ -187,53 +187,96 @@ async def test_start_game_called_on_match(
...
+    full_map_pool = mock.Mock()
+    queue.add_map_pool(newbie_map_pool, None, 500)
+    queue.add_map_pool(full_map_pool, 500, None)
 
-    chosen_map = await ladder_service.choose_map([None])
+    await ladder_service.start_game(p1, p2)
 
-    assert chosen_map is not None
+    newbie_map_pool.choose_map.assert_called_once()
+    full_map_pool.choose_map.assert_not_called()
```
Comment (norraxx) [inline tests/unit_tests/test_ladder_service.py:237]: `test_start_game_map_selection` and `test_start_game_map_selection_newbies` could be one test. Use `@pytest.mark.parametrize` to configure `player_factory`s...

Comment (norraxx) [reply tests/unit_tests/test_ladder_service.py:237]: OR even more, if you extract body:
```python
p1 = player_factory(
        ladder_rating=(2000, 50),
        ladder_games=1000,
    )
    p2 = player_factory(
        ladder_rating=(1500, 100),
        ladder_games=1000,
    )

    queue = ladder_service.queues["ladder1v1"]
    queue.map_pools.clear()
    newbie_map_pool = mock.Mock()
    full_map_pool = mock.Mock()
    queue.add_map_pool(newbie_map_pool, None, 500)
    queue.add_map_pool(full_map_pool, 500, None)

    await ladder_service.start_game(p1, p2)
```
You can make your test on 7 rows + 3 parametrize, 1 arguments, 1 call common method, 2 assert calls

But don't miss with assert calls on `map_pools`

PS: this comment touches also `test_start_game_map_selection_pros`

File: tests/unit_tests/test_map_pool.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    chosen_map = map_pool.choose_map([4, 5, 6])
+
+    assert chosen_map is not None
+    assert chosen_map in maps
+
+
+def test_choose_map_all_maps_played_returns_least_played(map_pool_factory):
+    # Large enough so the test is unlikely to pass by chance
+    NUM_MAPS = 1000
+    LIMIT = 3
```
Comment (norraxx) [inline tests/unit_tests/test_map_pool.py:68]: `Variable in function should be lowercase`

File: tests/unit_tests/test_map_pool.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    random.shuffle(maps)
+    map_pool = map_pool_factory(maps=maps)
+
+    chosen_map = map_pool.choose_map(played_map_ids)
+
+    # Map 1 was played only once
+    assert chosen_map == (1, "some_map", "maps/some_map.v001.zip")
+
+
+@given(history=st.lists(st.integers()))
```
Comment (norraxx) [inline tests/unit_tests/test_map_pool.py:89]: `O_o` WOW! something new. Don't get how it works, can You explain it to me, please? :-)

File: tests/unit_tests/test_map_pool.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    ])
+
+    # Make the probability very low that the test passes because we got lucky
+    for _ in range(20):
+        chosen_map = map_pool.choose_map(history)
+        assert chosen_map == (1, "CHOOSE_ME", "maps/choose_me.v001.zip")
+
+
+def test_choose_map_raises_on_empty_map_pool(map_pool_factory):
+    map_pool = map_pool_factory()
```
Comment (norraxx) [inline tests/unit_tests/test_map_pool.py:102]: If `MapPool` will modify the `maps` as function param in `map_pool_factory` fixture, you'll fail on this test.
I'm offering to not to use `= []` as default param in fixture, better use `maps or []` as when it's needed.

File: tests/unit_tests/test_matchmaker_queue_algorithm.py [hunk 8]
```diff
@@ -41,8 +41,8 @@ def st_searches_list(draw, min_players=1, max_players=10, max_size=30):
     )
 
 
-@pytest.fixture
-def p(player_factory):
+@pytest.fixture(scope="session")
+def player_factory(player_factory):
```
Comment (norraxx) [inline tests/unit_tests/test_matchmaker_queue_algorithm.py:45]: :+1:

File: server/ladder_service.py [hunk 9]
```diff
@@ -37,19 +40,102 @@ def __init__(
...
+    async def update_data(self) -> None:
+        async with self._db.acquire() as conn:
+            # Legacy ladder1v1 map pool
+            result = await conn.execute(
+                "SELECT ladder_map.idmap, "
+                "table_map.name, "
+                "table_map.filename "
+                "FROM ladder_map "
+                "INNER JOIN table_map ON table_map.id = ladder_map.idmap"
+            )
```
[the remark this answers is not in the document]
Comment (cleborys) [reply server/ladder_service.py:73]: I have come to like the SQLAlchemy query builder quite a bit, could also use that.

File: server/matchmaker/map_pool.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        amount of times.
+        """
+        if not self.maps:
+            self._logger.critical(
+                "Trying to choose a map from an empty map pool: %s", self.name
+            )
+            raise RuntimeError(f"Map pool {self.name} not set!")
+
+        # Make sure the counter has every available map
+        counter = Counter(self.maps.keys())
```
Comment (cleborys) [inline server/matchmaker/map_pool.py:37]: If you wanted to avoid adding one to every count, you could use
```python
counter = Counter({map: 0 for map in self.maps.keys()})
```
which might make it more future-proof (but adds one dict-comp).

File: server/matchmaker/map_pool.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        counter.update(id_ for id_ in played_map_ids if id_ in self.maps)
+
+        least_common = counter.most_common()[::-1]
+        least_count = least_common[0][1]
+        # Find the index of one past the last map with lowest count
+        for i, (_, count) in enumerate(least_common):
+            if count != least_count:
+                break
+        else:
+            i += 1
```
[the remark this answers is not in the document]
Comment (cleborys) [reply server/matchmaker/map_pool.py:47]: Oh, you can probably just replace it by `i=len(least_common)`?

File: server/matchmaker/matchmaker_queue.py [hunk 12]
```diff
@@ -37,18 +38,37 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+    ):
+        self.map_pools[map_pool.id] = (map_pool, min_rating, max_rating)
+
+    def get_map_pool_for_rating(self, rating: int) -> Optional[MapPool]:
+        for map_pool, min_rating, max_rating in self.map_pools.values():
+            if min_rating is not None and rating < min_rating:
+                continue
+            if max_rating is not None and rating > max_rating:
+                continue
+            return map_pool
```
Comment (cleborys) [inline server/matchmaker/matchmaker_queue.py:71]: I guess there are some edge cases where what map pool you get depends on the order of `self.map_pools.values()`, but I think that is ok in the interest of being robust against people specifying rating bounds however they see fit.
e.g. if `(min_rating, max_rating)` is `[(1,2), (2,3)]` and your rating is exactly `2`.
Or of course if they overlap like `[(1,3),(2,3)]`.

File: server/matchmaker/matchmaker_queue.py [hunk 13]
```diff
@@ -104,7 +125,7 @@ async def search(self, search: Search) -> None:
         assert search is not None
 
         try:
-            with MatchmakerSearchTimer(self.queue_name):
+            with MatchmakerSearchTimer(self.name):
```
[the remark this answers is not in the document]
Comment (cleborys) [reply server/matchmaker/matchmaker_queue.py:128]: That would be my fault, but I don't completely understand. Could you elaborate, @norraxx ?
Are you saying that `self.start_time = time.monotonic()` in `MatchmakerSearchTimer.__enter__` might error?

File: server/ladder_service.py [hunk 14]
```diff
@@ -1,25 +1,28 @@
...
 from server.db import FAFDatabase
 from server.rating import RatingType
 from sqlalchemy import and_, func, select, text
 
 from .async_functions import gather_without_exceptions
-from .config import LADDER_ANTI_REPETITION_LIMIT
+from .config import LADDER_ANTI_REPETITION_LIMIT, NEWBIE_MIN_GAMES
 from .core import Service
 from .db.models import game_featuredMods, game_player_stats, game_stats
+from .db.models import map as t_map
```
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:15]: i'm offering to name with full name like `map_model` or `table_map`, shortenings are not giving others much information

File: server/ladder_service.py [hunk 15]
```diff
@@ -37,19 +40,102 @@ def __init__(
...
+                    map_version.c.map_id,
+                    map_version.c.filename,
+                    t_map.c.display_name
+                ]).select_from(
+                    map_pool.join(map_pool_map_version)
+                    .join(map_version)
+                    .join(t_map)
+                )
+            )
+            map_pool_maps = {}
```
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:93]: `Pair`? You mean `Tuple`?

File: server/ladder_service.py [hunk 16]
```diff
@@ -37,19 +40,102 @@ def __init__(
...
+                id_ = row[map_pool.c.id]
+                name = row[map_pool.c.name]
+                if id_ not in map_pool_maps:
+                    map_pool_maps[id_] = (name, list())
+                _, map_list = map_pool_maps[id_]
+                map_list.append(
+                    Map(
+                        row[map_version.c.map_id],
+                        row[t_map.c.display_name],
+                        row[map_version.c.filename]
```
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:104]: Check, if `row.map_id` works too, please

File: server/ladder_service.py [hunk 17]
```diff
@@ -168,7 +254,20 @@ async def start_game(self, host: Player, guest: Player):
...
+                "ladder1v1"
+            )
+            rating = min(
+                player.ratings[RatingType.LADDER_1V1][0]
+                if player.game_count[RatingType.LADDER_1V1] > NEWBIE_MIN_GAMES
+                else 0
+                for player in (host, guest)
+            )
+            map_pool = self.queues["ladder1v1"].get_map_pool_for_rating(rating)
+            if not map_pool:
```
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:268]: no, need to check for None ;-) in such conditions it's easier to check for result, is there? fine. no result - ok, we continue, no need for none, if it's not needed

File: server/matchmaker/matchmaker_queue.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (norraxx) [reply server/matchmaker/matchmaker_queue.py:128]: OMG, my fault, Ignore it, I had changes in that file, that i didn't notice. Sorry.

File: server/db/models.py [hunk 18]
```diff
@@ -176,6 +176,68 @@
...
+    'matchmaker_queue_map_pool', metadata,
+    Column('matchmaker_queue_id',   Integer,    ForeignKey('matchmaker_queue.id'),  nullable=False),
+    Column('map_pool_id',           Integer,    ForeignKey('map_pool.id'),          nullable=False),
+    Column('min_rating',            Integer),
+    Column('max_rating',            Integer),
+)
+
 moderation_report = Table(
    'moderation_report', metadata,
    Column('id',                     Integer,                        primary_key=True),
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/db/models.py:243]: Hmm, I'm trying to keep the python object names consistent with the mysql table name

File: server/ladder_service.py [hunk 19]
```diff
@@ -1,25 +1,28 @@
...
 import contextlib
-import random
 from collections import defaultdict
-from typing import Dict, List, NamedTuple, Set
+from typing import Dict, List, Set
 
+import aiocron
 from server.db import FAFDatabase
 from server.rating import RatingType
 from sqlalchemy import and_, func, select, text
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:9]: What are you using to sort imports? I'm using `isort` currently.

File: server/ladder_service.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:15]: The worst part is that `table_map` is not even a table, it's a view.

File: server/ladder_service.py [hunk 20]
```diff
@@ -37,19 +40,102 @@ def __init__(
...
 
     async def initialize(self) -> None:
+        await self.update_data()
+        self._update_cron = aiocron.crontab('*/10 * * * *', func=self.update_data)
         await asyncio.gather(*[
             queue.initialize() for queue in self.queues.values()
         ])
         asyncio.create_task(self.handle_queue_matches())
 
+    async def update_data(self) -> None:
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:64]: `update_data` is a standard function that is present on many services. It always does the same thing which is to pull in and cache some data from the database.

File: server/ladder_service.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:73]: Is it worth it though at this point? This query is just going to get removed as soon as we've transitioned to the new tables. And then presumabley the old tables/views can be dropped.

File: server/ladder_service.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:104]: Yes it seems to work. The docs are not very good https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.RowProxy but they do mention that it's a named tuple.

File: server/ladder_service.py [hunk 21]
```diff
@@ -37,19 +40,102 @@ def __init__(
...
+                name = row[matchmaker_queue.c.technical_name]
+                if name not in self.queues:
+                    self.queues[name] = MatchmakerQueue(
+                        name=name,
+                        name_key=row[matchmaker_queue.c.name_key],
+                        game_service=self.game_service
+                    )
+                queue = self.queues[name]
+                if name not in queue_names:
+                    queue.map_pools.clear()
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:125]: I need to avoid destroying MatchmakerQueues to make sure the legacy one that's hardcoded doesn't get destroyed until there is actually a replacement for it in the database.

File: server/ladder_service.py [hunk 22]
```diff
@@ -37,19 +40,102 @@ def __init__(
...
+                map_pool_id = row[matchmaker_queue_map_pool.c.map_pool_id]
+                map_pool_name, map_list = map_pool_maps[map_pool_id]
+                queue.add_map_pool(
+                    MapPool(map_pool_id, map_pool_name, map_list),
+                    row[matchmaker_queue_map_pool.c.min_rating],
+                    row[matchmaker_queue_map_pool.c.max_rating]
+                )
+                queue_names.add(name)
+            # Remove queues that don't exist anymore
+            for queue_name in list(self.queues.keys()):
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:135]: :P

File: server/ladder_service.py [hunk 23]
```diff
@@ -168,7 +254,20 @@ async def start_game(self, host: Player, guest: Player):
             host.state = PlayerState.HOSTING
             guest.state = PlayerState.JOINING
 
-            (map_id, map_name, map_path) = await self.choose_map([host, guest])
+            played_map_ids = await self.get_game_history(
+                [host, guest],
+                "ladder1v1"
+            )
+            rating = min(
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:261]: Ah, need to update the description then. I convinced myself that lowest was actually better.

File: server/ladder_service.py [hunk 24]
```diff
@@ -168,7 +254,20 @@ async def start_game(self, host: Player, guest: Player):
...
+                else 0
+                for player in (host, guest)
+            )
+            map_pool = self.queues["ladder1v1"].get_map_pool_for_rating(rating)
+            if not map_pool:
+                raise RuntimeError(f"No map pool available for rating {rating}!")
+            (map_id, map_name, map_path) = map_pool.choose_map(played_map_ids)
 
             game = self.game_service.create_game(
                 game_mode='ladder1v1',
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:273]: Yes `game_mode` means featured mod

File: server/matchmaker/map_pool.py [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+from ..types import Map
+
+
+@with_logger
+class MapPool(object):
+    def __init__(
+        self,
+        map_pool_id: int,
+        name: str,
+        maps: List[Map] = []
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/matchmaker/map_pool.py:15]: Ah, that's clever

File: server/matchmaker/map_pool.py [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+    ):
+        self.id = map_pool_id
+        self.name = name
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/matchmaker/map_pool.py:18]: ??? Typing is already defined by the method signature

File: server/matchmaker/map_pool.py (hunk 10 above)
Comment (Askaholic) [reply server/matchmaker/map_pool.py:37]: I did that at first, but then I realized that I don't care what the actual count is. Feels nicer to me to have less code and no intermediate data structure.

File: server/matchmaker/map_pool.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/matchmaker/map_pool.py:47]: No, because `least_common` is all of the maps, but sorted in reverse order by number of occurrences. As the comment says, we need this code in order to trim off all the maps which have higher play counts.

File: tests/unit_tests/test_ladder_service.py [hunk 27]
```diff
@@ -187,53 +187,96 @@ async def test_start_game_called_on_match(
...
-        assert chosen_map == (4, "CHOOSE_ME", "maps/choose_me.v001.zip")
+    newbie_map_pool.choose_map.assert_called_once()
+    full_map_pool.choose_map.assert_not_called()
 
 
-async def test_choose_map_all_maps_played(ladder_service: LadderService):
-    ladder_service.get_ladder_history = CoroutineMock(
-        return_value=[1, 2, 3]
+async def test_start_game_map_selection_newbies(
+    ladder_service: LadderService, player_factory
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply tests/unit_tests/test_ladder_service.py:216]: `player_factory` is a callable, typing for it is annoying :P

File: tests/unit_tests/test_map_pool.py (hunk 6 above)
Comment (Askaholic) [reply tests/unit_tests/test_map_pool.py:89]: `hypothesis` is a fuzzer. It means you tell it the type of data that is valid for your test, and it will give you randomly generated data. Often times this can help you find a lot of edge cases in your code that you didn't think about before. Usually I find that it immediately fails on `[]` empty input or something really simple that I forgot to think about.

It actually is more fancy than just random data. It tries to find minimal failing examples so it's as easy as possible to debug why that data made it fail.

File: server/ladder_service.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (cleborys) [reply server/ladder_service.py:9]: Oh, that's a good thing to know :)
If you're open to it I would like to have such things (mostly this and `flake8`) as pre-commit hooks for everyone.

File: server/ladder_service.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (cleborys) [reply server/ladder_service.py:93]: Yup

File: server/matchmaker/map_pool.py (hunk 10 above)
Comment (cleborys) [reply server/matchmaker/map_pool.py:37]: Makes sense!

File: server/matchmaker/map_pool.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (cleborys) [reply server/matchmaker/map_pool.py:47]: Maybe I'm mistaken about the flow of execution, but I thought that the problem was potentially hitting `i += 1` when `least_common` is empty, since then `i` is not defined.
And as you only hit the `else` clause if you don't break, i.e. if every `count` is equal to `least_count` you will only ever hit the `i += 1` line with either `i` undefined or `i == len(least_common) - 1` after iterating through all of `least_common`, so might as well replace `i += 1` with `i = len(least_common)`?

Anyways I don't think it matters much with the empty check before, it's just to please whatever linter noraxx was using.

File: server/ladder_service.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (cleborys) [reply server/ladder_service.py:15]:
```python
import table_map as thing_called_table_map_which_is_actually_a_view
```
Self-documenting code best-practices for the win!

File: server/ladder_service.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:9]: Pycharm :-)

File: server/ladder_service.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:15]: > The worst part is that `table_map` is not even a table, it's a view.

Should i migrate it to table?

File: server/ladder_service.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:73]: Maybe make issue, and add todo + issue number

File: server/ladder_service.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:104]: btw, you can convert it namedtuple to dict easy and it's easy to dump to json also

File: server/ladder_service.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (norraxx) [reply server/ladder_service.py:273]: maybe let's have some constants, that will have proper naming, one for mode, one for RatingType (already exists)

File: tests/unit_tests/test_map_pool.py (hunk 6 above)
Comment (norraxx) [reply tests/unit_tests/test_map_pool.py:89]: I'll need to check this at my work, this is very interesting thing.

File: server/ladder_service.py [hunk 28]
```diff
@@ -37,19 +40,102 @@ def __init__(
...
+                'ladder1v1',
+                map_pools=[(self.ladder_1v1_map_pool, None, None)]
+            )
         }
 
         self.searches: Dict[str, Dict[Player, Search]] = defaultdict(dict)
 
     async def initialize(self) -> None:
+        await self.update_data()
+        self._update_cron = aiocron.crontab('*/10 * * * *', func=self.update_data)
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:58]: Not that it is ever used. It just has to be saved somewhere so it doesn't get garbage collected.

File: server/ladder_service.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:135]: Actually it is needed. See this example:
```python3

>>> d = dict(a=10, b=13, c=39, d=123, e=123, f=120)
>>> for key in d.keys():
...     del d[key]
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dictionary changed size during iteration
```

File: server/matchmaker/map_pool.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/matchmaker/map_pool.py:47]: Oh I see what you mean

File: server/matchmaker/matchmaker_queue.py [hunk 29]
```diff
@@ -37,18 +38,37 @@ def __exit__(self, exc_type, exc_value, traceback):
...
+        game_service: "GameService",
+        name: str,
+        name_key: str,
+        map_pools: List[Tuple[MapPool, Optional[int], Optional[int]]] = []
     ):
         self.game_service = game_service
-        self.queue_name = queue_name
+        self.name = name
+        self.name_key = name_key
+        self.map_pools = {info[0].id: info for info in map_pools}
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/matchmaker/matchmaker_queue.py:49]: Don't need to. Mypy can figure it out

File: server/ladder_service.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/ladder_service.py:14]: I don't think so because this PR makes it obsolete anyways. Better make a PR that drops the view entirely so we can merge it when we're ready.

File: server/db/models.py [hunk 30]
```diff
@@ -176,6 +176,68 @@
     Column('expires_at',    DateTime)
 )
 
+map = Table(
```
[the remark this answers is not in the document]
Comment (Askaholic) [reply server/db/models.py:179]: I'm going to leave it as is for consistency with the other models. Models should have the same name as the table they're representing. I think it is clear that `db.map` is not the same as builtins `map`, and when you import it you can rename it as you please.

[the remark this answers is not in the document]
Comment (norraxx) [reply server/db/models.py:179]: np

File: tests/unit_tests/test_map_pool.py [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+    assert chosen_map in maps
+
+
+def test_choose_map_all_maps_played_returns_least_played(map_pool_factory):
+    # Large enough so the test is unlikely to pass by chance
+    num_maps = 1000
+    limit = 3
+
+    played_map_ids = [
+        # Set up player history so map 1 is played the most
```
Comment (cleborys) [inline tests/unit_tests/test_map_pool.py:71]: I think you mean "played the least"?

File: tests/unit_tests/test_matchmaker_queue.py [hunk 32]
```diff
@@ -224,7 +225,83 @@ async def test_queue_pop_time_moving_average_size():
...
+    queue = queue_factory()
+    map_pool1 = MapPool(0, "pool1")
+    map_pool2 = MapPool(1, "pool2")
+    queue.add_map_pool(map_pool1, low1, high1)
+    queue.add_map_pool(map_pool2, low2, high2)
+
+    if low1 <= rating <= high1:
+        assert queue.get_map_pool_for_rating(rating) is map_pool1
+    elif low2 <= rating <= high2:
+        assert queue.get_map_pool_for_rating(rating) is map_pool2
```
Comment (cleborys) [inline tests/unit_tests/test_matchmaker_queue.py:298]: I think this test relies on `queue` saving the `MapPool`s in  order added and returning the first pool that was added if there is ambiguity.
That's not a bad thing, but maybe something to keep in mind - this test might break if `queue` ever chooses to store map pools in e.g. a dict and then we might not have wanted to document this assumption.
I'm ok with it as is but you could consider
```python3
eligible_map_pools = [map_pools[i] for i in range(2) if low[i] <= rating <= high[i]]
if eligible_map_pools:
  assert queue.get_map_pool_for_rating(rating) in eligible_map_pools
else:
  assert queue.get_map_pool_for_rating(rating) is None
```

File: tests/unit_tests/test_map_pool.py (hunk 31 above)
Comment (Askaholic) [reply tests/unit_tests/test_map_pool.py:71]: Those pesky comments. Always wrong :P
